### ISSUE TITLE
ParseException.java is always public

### ISSUE THREAD

**Issue by mtf90** [javacc/javacc#277]
Despite the documentation on [SUPPORT_CLASS_VISIBILITY_PUBLIC](https://javacc.github.io/javacc/documentation/grammar.html#option-binding), the generated `ParseException.java` class is always public. As seen in the [ParseException.template](https://github.com/javacc/javacc/blob/master/src/main/resources/templates/ParseException.template), there is no `${SUPPORT_CLASS_VISIBILITY_PUBLIC?public :}` check like in other templates.

I encountered this issue while enabling the `javadoc` linter, since the constructors have missing `@param` descriptors in their documentation. While this may be a separate issue, I think the class visibility should still be consistent.

**Comment by MarcMazas** [comment]
Well ...
1 - the documentation states that if the option is set to false the classes will be generated to private, which seems not to be true - it will have the default no modifier (package-private).
2 - I think that people how introduced this option thought that it would be needed for the support classes that are not usually directly used in the parsers' callers (like CharStream or TokenManager), but not for the more often used ParseException class (think : can one really write a tool T in package a that calls a parser P in package b without having to catch b.PE in a.T ?) : 
3 - IMHO this option is not very interesting : if one wants to fine-grained control the generated classes, he should switch to javaCC 8 and customize the templates.
4 - for javacc 7 we'll have to fix either only the docs or the docs + the PE template.

**Comment by mtf90** [comment]
Thanks for the quick response. To your points:

1. I think the documentation is fine. It states
   >  If set to `false`, the classes will be generated with package `private` visibility.

   The only misleading thing here may be the highlighting of the `private` keyword, since JavaCC does not actually generate a `private class` but a "package-private" `class` (i.e., the lack of all visibility modifiers).
2. The main appeal of this option (at least to me) is to hide the implementation details of the parser. So typically, I would just wrap the `ParseException` in a custom execption type and expose this type in the public interfaces. Especially if you have multiple grammars, you would have several (identical) `ParseException`s which clutter the API/documentation.
3. Can you maybe point me in the correct direction? Both [javacc](https://github.com/javacc/javacc) and [javacc-8](https://github.com/javacc/javacc-8) seem to refer to the same artifacts. Although, to be honest, maintaining a custom template for a class I will probably never touch anyway seems to introduce an unnecessary maintenance burden.

**Comment by zbynek** [comment]
ad 3) javacc 8 is consists of multiple Maven artifacts (the core + one generator for each output language), each has its own GitHub repo. The ParseException template is in https://github.com/javacc/javacc-8-java/blob/release/src/main/resources/templates/ParseException.template

**Comment by mtf90** [comment]
Dear @zbynek, thanks for the heads-up. Unfortunately, I use the `javacc-maven-plugin` which [does not seem to support](https://github.com/mojohaus/javacc-maven-plugin/pull/4) the new javacc-8 version yet. But even if it did, both [javacc](https://github.com/javacc/javacc/blob/master/src/main/java/org/javacc/utils/OutputFileGenerator.java#L65) and [javacc-8](https://github.com/javacc/javacc-8-core/blob/release/src/main/java/org/javacc/utils/TemplateBuilder.java#L68) seem to load the templates as resources. So unless I hijack the classloader to inject my own template, I don't see a (convenient) way for providing custom templates, or am I missing something?

While I was able to workaround my initial problem, I still think the proper way to address this issue is to include the `public` check in the `ParseException.template` as well, just like in all the other templates. Given the documentation, I think this was the intention from the beginning and it just slipped through. I'm also happy to provide a PR, if you are willing to merge it.

**Comment by zbynek** [comment]
I'm not sure if `ParseException` should count as "support class" though -- in most usecases `ParseException` will be used in the signature of public methods of the generated parser, and should therefore be also public. So maybe a new option is needed?

Btw. you linked this issue from a PR that's about Javadoc issues -- does that mean that the current `ParseException` is giving you some Javadoc warnings?

**Comment by mtf90** [comment]
> I'm not sure if `ParseException` should count as "support class" though -- in most usecases `ParseException` will be used in the signature of public methods of the generated parser, and should therefore be also public. So maybe a new option is needed?

I think I see the issue. In my use-case, I want the parser to be hidden from the public API which is why my parser definition (after `PARSER_BEGIN`) is package-private and I use `SUPPORT_CLASS_VISIBILITY_PUBLIC = false;`. However, you are concerned with the case of a public parser definition + `SUPPORT_CLASS_VISIBILITY_PUBLIC = false;`. In this case, the `ParseException` should probably be `public`, indeed.

Could this then probably be auto-detected upon parsing the parser definition? If the parser itself is package-private, even its public methods cannot be accessed outside of its package, so a package-private `ParseException` should be safe (and is probably intended by the grammar author). If the parser is public, the `ParseException` should be public as well, irrespective of the  `SUPPORT_CLASS_VISIBILITY_PUBLIC` value (as it is now).

> Btw. you linked this issue from a PR that's about Javadoc issues -- does that mean that the current `ParseException` is giving you some Javadoc warnings?

Yes, doclint complained:

```
Warning:  Javadoc Warnings
Warning:  /home/runner/work/automatalib/automatalib/modelchecking/ltsmin/target/generated-sources/annotations/net/automatalib/modelchecker/ltsmin/ParseException.java:34: warning: no @param for currentTokenVal
Warning:  public ParseException(Token currentTokenVal,
Warning:  ^
Warning:  /home/runner/work/automatalib/automatalib/modelchecking/ltsmin/target/generated-sources/annotations/net/automatalib/modelchecker/ltsmin/ParseException.java:34: warning: no @param for expectedTokenSequencesVal
Warning:  public ParseException(Token currentTokenVal,
Warning:  ^
Warning:  /home/runner/work/automatalib/automatalib/modelchecking/ltsmin/target/generated-sources/annotations/net/automatalib/modelchecker/ltsmin/ParseException.java:34: warning: no @param for tokenImageVal
Warning:  public ParseException(Token currentTokenVal,
Warning:  ^
Warning:  /home/runner/work/automatalib/automatalib/modelchecking/ltsmin/target/generated-sources/annotations/net/automatalib/modelchecker/ltsmin/ParseException.java:60: warning: no @param for message
Warning:  public ParseException(String message) {
Warning:  ^
```
However, this was on JDK <17, so only incomplete docs are flagged. JDK >= 17 (and 21 most definitely) also flags *any* missing documentation on public members.

**Comment by MarcMazas** [comment]
Ok, thanks. I note:
- special processing for public / none modifier for ParserException: take the parser's class's one; not a "support" class
- add decent javadoc comments for all public fields / methods: but what about the generated productions methods (marked final public): will they also need a javadoc comment?
- clarify how to customize templates

By the way, in a close area, what do you think about the GENERATE_ANNOTATIONS option, which adds a @SuppressWarnings("all")? Should it be also used in all support class & ParseException? Knowing that TokenManager class is generated with @SuppressWarnings("unused").

**Comment by zbynek** [comment]
@MarcMazas 

> add decent javadoc comments for all public fields / methods: but what about the generated productions methods (marked final public): will they also need a javadoc comment?

https://github.com/javacc/javacc-8-java/pull/4/files provides enough Javadoc to the generated classes, so that default checks should not fail -- a Javadoc check for the generated code was also added to CI to verify that. Whether that Javadoc counts as *decent* is a question, but it's a starting point and I'm happy to improve it if there's some feedback.

I'm not sure if Javadoc compliance is important enough for a 7.0 backport.

@mtf90 
> both [javacc](https://github.com/javacc/javacc/blob/master/src/main/java/org/javacc/utils/OutputFileGenerator.java#L65) and [javacc-8](https://github.com/javacc/javacc-8-core/blob/release/src/main/java/org/javacc/utils/TemplateBuilder.java#L68) seem to load the templates as resources. So unless I hijack the classloader to inject my own template, I don't see a (convenient) way for providing custom templates

If you're using Maven or Gradle plugins, you can just add the generated `ParseException` (with your improvements) to git in the right package under `src/main/java`, then the plugin won't generate another one. (Actually it generates it into temp folder, but then [doesn't copy it to your classpath](https://github.com/mojohaus/javacc-maven-plugin/commit/89c1f574c39f95ec904f1250714e18f84acd6c11)). Keeping a generated file in Git is not great, but it's the same as keeping the template (there is just 1 statement in the template that is optionally included, depending on a parameter).

**Comment by mtf90** [comment]
Thanks @zbynek, your advice does work and allows me to provide a custom (package-private) `ParseException`.

Do you consider this issue as resolved then? While the ability to provide a custom implementation allows for a workaround and somewhat lessens the impact of the original issue, I think the downside of having to maintain this file as an actual source file still justifies thinking about the ability to adjust the generation process (either by means of auto detection of another config property).

Any opinions on this?

**Comment by zbynek** [comment]
@mtf90 I don't think this is really resolved, my PR only addresses the second suggestion by @MarcMazas . It might still be useful to implement the first one (though I'd prefer to do all non-critical fixes only in JavaCC 8 and make them available to Maven users by cleaning up https://github.com/mojohaus/javacc-maven-plugin/pull/4).

**Comment by MarcMazas** [comment]
Just tested on javacc 7, under an ant script: in your product's project:

- create a specific resource directory: eg: s**rc/main/javacc-resources** and a sub-directory **templates**
- copy under **templates** any template you want: eg **ParserException.template** and modify it
- set your javacc generation java task with such a classpath: **src/main/javacc-resources:<path-to-javacc-7.0.n.jar>**
- do not set the **JAVA_TEMPLATE_TYPE** option to "**modern**", keep the "**classic**" default

and you should get the corresponding support file generated from your template.

I let somebody find if there is a direct solution for maven (scope provided or system for the different javacc-maven-plugin (s)?).
